judgels-backends/uriel/uriel-app/src/main/java/judgels/uriel/contest/scoreboard/icpc/IcpcScoreboardProcessor.java

```diff
@@ -22,6 +35,133 @@ public Scoreboard parseFromString(ObjectMapper mapper, String json) {
         }
     }
 
+    @Override
+    public String computeToString(
+            ObjectMapper mapper,
+            ScoreboardState scoreboardState,
+            Contest contest,
+            ContestModulesConfig contestModulesConfig,
+            Map<String, Optional<Instant>> contestantStartTimesMap,
+            List<Submission> submissionList) {
+        if (!contestModulesConfig.getIcpcStyle().isPresent()) {
+            throw new RuntimeException("ICPC-style contest given but IcpcStyleConfig not found");
+        }
+        IcpcStyleModuleConfig icpcStyleModuleConfig = contestModulesConfig.getIcpcStyle().get();
+
+        List<String> problemJids = scoreboardState.getProblemJids();
+        Set<String> problemJidsSet = new HashSet<>(problemJids);
+        Set<String> contestantJids = scoreboardState.getContestantJids();
+
+        Map<String, List<Submission>> submissionsMap = new HashMap<>();
+        List<Submission> mutableSubmissionList = new ArrayList<>(submissionList);
+        mutableSubmissionList.sort(Comparator.comparing(Submission::getTime));
+        mutableSubmissionList.forEach(s -> {
+            submissionsMap.putIfAbsent(s.getUserJid(), new ArrayList<>());
+            submissionsMap.get(s.getUserJid()).add(s);
+        });
+
+        Map<String, String> firstSolveId = new HashMap<>();
+        mutableSubmissionList.stream()
+                .filter(s -> s.getLatestGrading().isPresent()
+                        && s.getLatestGrading().get().getVerdict().equals(Submission.ACCEPTED)
+                        && contestantJids.contains(s.getUserJid()))
+                .forEach(s -> firstSolveId.putIfAbsent(s.getProblemJid(), s.getJid()));
+
+        List<IcpcScoreboardEntry> scoreboardEntryList = new ArrayList<>();
+        for (String contestantJid : submissionsMap.keySet()) {
+            List<Submission> submissions = submissionsMap.get(contestantJid);
+            long lastAcceptedPenalty = 0;
+            Map<String, Integer> attemptsMap = new HashMap<>();
+            problemJids.forEach(p -> attemptsMap.put(p, 0));
+            Map<String, Long> penaltyMap = new HashMap<>();
+            problemJids.forEach(p -> penaltyMap.put(p, 0L));
+            Map<String, IcpcScoreboardProblemState> stateMap = new HashMap<>();
```

Comment:
```suggestion
            Map<String, IcpcScoreboardProblemState> problemStateMap = new HashMap<>();
```

(because we also have scoreboard state)